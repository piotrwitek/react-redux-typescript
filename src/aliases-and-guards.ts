/**
 * Primitive
 * @desc Type representing primitive types in TypeScript: `number | bigint | boolean | string | symbol`
 * @example
 *   type Various = number | string | object;
 *
 *    // Expect: object
 *   type Cleaned = Exclude<Various, Primitive>
 */
export type Primitive = number | bigint | boolean | string | symbol;

/**
 * Falsy
 * @desc Type representing falsy values in TypeScript: `null | undefined | false | 0 | ''`
 * @example
 *   type Various = 'a' | 'b' | undefined | false;
 *
 *   // Expect: "a" | "b"
 *   Exclude<Various, Falsy>;
 */
export type Falsy = null | undefined | false | 0 | '';

/**
 * Tests for one of the [`Primitive`](https://developer.mozilla.org/en-US/docs/Glossary/Primitive) types using the JavaScript [`typeof`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof) operator
 *
 * Clarification: TypeScript overloads this operator to produce TypeScript types if used in context of types.
 *
 * @param val The value to be tested
 * @returns If `val` is primitive. If used in the flow of the program typescript will infer type-information from this.
 *
 * @example
 *   const consumer = (value: Primitive | Primitive[]) => {
 *       if (isPrimitive(value)) {
 *           return console.log('Primitive value: ', value);
 *       }
 *       // type of value now inferred as Primitive[]
 *       value.map((primitive) => consumer(primitive));
 *   };
 */
export const isPrimitive = (val: unknown): val is Primitive => {
  switch (typeof val) {
    case 'string':
    case 'symbol':
    case 'number':
    case 'bigint':
    case 'boolean': {
      return true;
    }
    default:
      return false;
  }
};

/**
 * Tests for Falsy by simply applying negation `!` to the tested `val`.
 *
 * The value is mostly in added type-information and explicity,
 * but in case of this simple type much the same can often be archived by just using negation `!`:
 * @example
 *   const consumer = (value: boolean | Falsy) => {
 *     if (!value) {
 *         return ;
 *     }
 *     type newType = typeof value; // === true
 *     // do stuff
 *   };
 */
export const isFalsy = (val: unknown): val is Falsy => !val;
