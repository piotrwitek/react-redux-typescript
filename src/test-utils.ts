export function testType<T>(a: T): T { return undefined as any; }

// export interface IsType<T> {
//   (v: T): '1';
//   (v: any): '0';
// }
