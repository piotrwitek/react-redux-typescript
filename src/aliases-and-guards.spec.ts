import { testType } from '../utils/test-utils';
import { Primitive, isPrimitive, Falsy, isFalsy } from './aliases-and-guards';

// @dts-jest:group Primitive
{
  // @dts-jest:pass:snap
  testType<Primitive>();
}

// @dts-jest:group isPrimitive
it('narrows to correct type', () => {
  const consumer = (param: Primitive[] | Primitive): string => {
    if (isPrimitive(param)) {
      // @dts-jest:pass:snap
      param;
      return String(param) + ' was Primitive';
    }
    // @dts-jest:pass:snap
    param;
    const resultArray = param
      .map(consumer)
      .map(rootString => '\n\t' + rootString);
    return resultArray.reduce((comm, newV) => comm + newV, 'this was nested:');
  };
});

// @dts-jest:group Falsy
{
  // @dts-jest:pass:snap
  testType<Falsy>();
}

// @dts-jest:group isFalsy-Falsy
it('returns true for falsy and narrows type', () => {
  const falsyTestVals: unknown[] = ['', null, undefined, false, 0];

  falsyTestVals.forEach(val => {
    if (isFalsy(val)) {
      // @dts-jest:pass:snap
      val;
    }
    // @dts-jest:pass:snap
    val;
  });

  const testResults = falsyTestVals.map(isFalsy);

  testResults.forEach(val => expect(val).toBe(true));
});

// @dts-jest:group isFalsy-Truthy
it('returns false for truthy and narrows type', () => {
  const truthyTestVals: unknown[] = [' ', true, {}, []];

  truthyTestVals.forEach(val => {
    if (isFalsy(val)) {
      // @dts-jest:pass:snap
      val;
    }
    // @dts-jest:pass:snap
    val;
  });

  const testResults = truthyTestVals.map(isFalsy);

  testResults.forEach(val => expect(val).toBe(false));
});
