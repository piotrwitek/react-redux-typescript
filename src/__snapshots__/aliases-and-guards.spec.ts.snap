// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Falsy testType<Falsy>() 1`] = `"Falsy"`;

exports[`Primitive testType<Primitive>() 1`] = `"Primitive"`;

exports[`isFalsy-Falsy val 1`] = `"Falsy"`;

exports[`isFalsy-Falsy val 2`] = `"unknown"`;

exports[`isFalsy-Truthy val 1`] = `"Falsy"`;

exports[`isFalsy-Truthy val 2`] = `"unknown"`;

exports[`isPrimitive param 1`] = `"Primitive"`;

exports[`isPrimitive param 2`] = `"Primitive[]"`;
